python/test/conftest.py

```diff
@@ -136,3 +139,64 @@ def tempdir(request):
 
     """
     return _create_tempdir(request)
+
+
+# Basic Conjugate Gradient solver
+def _cg(comm, A, b, x):
+    kmax = 500
+    rtol2 = 10 * np.finfo(x.array.dtype).eps
+
+    def _global_dot(comm, v0, v1):
+        return comm.allreduce(np.dot(v0, v1), MPI.SUM)
+
+    A_op = A.to_scipy()
+    nr = A_op.shape[0]
+    assert nr == A.index_map(0).size_local
+
+    # Create larger ghosted vector based on matrix column space
+    # and get initial y = A.x
+    p = dolfinx_vector(A.index_map(1), dtype=x.array.dtype)
+    p.array[:nr] = x.array[:nr]
+    p.scatter_forward()
+    y = A_op @ p.array
+
+    # Copy residual to p
+    r = b.array[:nr] - y
+    p.array[:nr] = r
+
+    # Iterations of CG
+    rnorm0 = _global_dot(comm, r, r)
+    rnorm = rnorm0
+    k = 0
+    while (k < kmax):
+        k += 1
+        p.scatter_forward()
+        y = A_op @ p.array
+        alpha = rnorm / _global_dot(comm, p.array[:nr], y)
+        x.array[:nr] += alpha * p.array[:nr]
+        r -= alpha * y
+        rnorm_new = _global_dot(comm, r, r)
+        beta = rnorm_new / rnorm
+        rnorm = rnorm_new
+        if (rnorm / rnorm0 < rtol2):
+            return
+        p.array[:nr] = beta * p.array[:nr] + r
+
+    raise RuntimeError(f"Solver exceeded max iterations ({kmax}).")
+
+
+@pytest.fixture(scope="function")
+def solver():
```

Comment:
Name is very generic - parallel case only work for SPD problems.

Reply:
Maybe I should always use CG, even in serial? At least it is consistent then... and I can change the name to `cg_solver`?

Reply:
It's probably better to use the same solver in serial and parallel.

I suggest adding [BiCGSTAB](https://en.wikipedia.org/wiki/Biconjugate_gradient_stabilized_method) for non-symmetric and indefinite problems.

Reply:
OK, I will change this to `cg_solver()` - maybe we can save BiCGStab for another PR.